package/src/lock.rs

```diff
@@ -0,0 +1,267 @@
+//! Lock files and lock file utilities
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    num::ParseIntError,
+    path::{Path, PathBuf},
+};
+
+use gix::ObjectId;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::{Error, IoResultExt},
+    snapshot::Snapshot,
+    Dependency, GitDependency, ManifestFile, Precise,
+};
+
+/// We need to give names to entries in the lock-file, so that we can refer to dependencies.
+///
+/// It's tempting to try to use `Precise` for identifying dependencies, but we can't because
+/// the lock file doesn't store paths on the local filesystem. So we identify packages by
+/// the name that its importer uses, and add numbers to ensure uniqueness.
+#[derive(
+    Clone,
+    Debug,
+    Default,
+    serde_with::SerializeDisplay,
+    serde_with::DeserializeFromStr,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+)]
+pub struct EntryName {
+    pub name: String,
+    pub id: u32,
+}
+
+impl std::fmt::Display for EntryName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.id == 0 {
+            write!(f, "{}", self.name)
+        } else {
+            write!(f, "{} {}", self.name, self.id)
+        }
+    }
+}
+
+impl std::str::FromStr for EntryName {
+    type Err = ParseIntError;
+
+    fn from_str(s: &str) -> Result<EntryName, ParseIntError> {
+        if let Some((name, num)) = s.split_once(' ') {
+            Ok(EntryName {
+                name: name.to_owned(),
+                id: num.parse()?,
+            })
+        } else {
+            Ok(EntryName {
+                name: s.to_owned(),
+                id: 0,
+            })
+        }
+    }
+}
+
+#[derive(Default)]
+struct LockFileNamer {
+    counts: HashMap<String, u32>,
+    assigned: HashMap<Precise, (String, u32)>,
+}
+
+impl LockFileNamer {
+    fn name(&mut self, name: &str, pkg: &Precise) -> EntryName {
+        if let Some((old_name, old_num)) = self.assigned.get(pkg) {
+            EntryName {
+                name: old_name.to_owned(),
+                id: *old_num,
+            }
+        } else {
+            let count = self.counts.get(name).copied().map_or(0u32, |i| i + 1);
+            self.counts.insert(name.to_owned(), count);
```

Comment:
I suspect this can be done in a one liner (and with only one hash lookup) with the `Entry` API.

Reply:
rustfmt disagrees :smile: But sure, I think the one-statement-over-five-lines version is clearer.